File: c++ version/201-300/207-course-schedule.cpp
bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
    vector<vector<int>> graph(numCourses, vector<int>());
    vector<int> indegree(numCourses, 0);
    for (const auto& pair : prerequisites) {
        graph[pair[1]].push_back(pair[0]);
        ++indegree[pair[0]];
    }

    queue<int> q;
    for (int i = 0; i < indegree.size(); ++i) {
        if (indegree[i] == 0) {
            q.push(i);
        }
    }
    while (!q.empty()) {
        int u = q.front();
        q.pop();
        for (auto v : graph[u]) {
            --indegree[v];
            if (indegree[v] == 0) {
                q.push(v);
            }
        }
    }

    for (int i = 0; i < indegree.size(); ++i) {
        if (indegree[i] != 0) {
            return false;
        }
    }

    return true;
}

File: python version/207-course-schedule.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        in_degree = [0 for _ in range(numCourses)]
        out_graph = {}
        for pair in prerequisites:
            in_degree[pair[0]] += 1
            if pair[1] not in out_graph:
                out_graph[pair[1]] = [pair[0]]
            else:
                out_graph[pair[1]].append(pair[0])
        queue = []

        for i in range(numCourses):
            if in_degree[i] == 0:
                queue.append(i)

        while queue:
            curr = queue.pop(0)
            if curr in out_graph:
                for each in out_graph[curr]:
                    in_degree[each] -= 1
                    if in_degree[each] == 0:
                        queue.append(each)
        
        for i in range(numCourses):
            if in_degree[i] != 0:
                return False
        return True